src/qr.rs

```diff
@@ -0,0 +1,469 @@
+const VCARD_SCHEME: &str = "BEGIN:VCARD";
+const SMTP_SCHEME: &str = "SMTP:";
+const HTTP_SCHEME: &str = "http://";
+const HTTPS_SCHEME: &str = "https://";
+
+// Make it easy to convert errors into the final `Lot`.
+impl Into<Lot> for Error {
+    fn into(self) -> Lot {
+        let mut l = Lot::new();
+        l.state = LotState::QrError;
+        l.text1 = Some(self.to_string());
+
+        l
+    }
+}
+
+/// Check a scanned QR code.
+/// The function should be called after a QR code is scanned.
+/// The function takes the raw text scanned and checks what can be done with it.
+pub fn check_qr(context: &Context, qr: impl AsRef<str>) -> Lot {
+    let qr = qr.as_ref();
+
+    info!(context, 0, "Scanned QR code: {}", qr);
+
+    if qr.starts_with(OPENPGP4FPR_SCHEME) {
+        decode_openpgp(context, qr)
+    } else if qr.starts_with(MAILTO_SCHEME) {
+        decode_mailto(context, qr)
+    } else if qr.starts_with(SMTP_SCHEME) {
+        decode_smtp(context, qr)
+    } else if qr.starts_with(MATMSG_SCHEME) {
+        decode_matmsg(context, qr)
+    } else if qr.starts_with(VCARD_SCHEME) {
+        decode_vcard(context, qr)
+    } else if qr.starts_with(HTTP_SCHEME) || qr.starts_with(HTTPS_SCHEME) {
+        Lot::from_url(qr)
+    } else {
+        Lot::from_text(qr)
+    }
+}
+
+/// scheme: `OPENPGP4FPR:FINGERPRINT#a=ADDR&n=NAME&i=INVITENUMBER&s=AUTH`
+///     or: `OPENPGP4FPR:FINGERPRINT#a=ADDR&g=GROUPNAME&x=GROUPID&i=INVITENUMBER&s=AUTH`
+fn decode_openpgp(context: &Context, qr: &str) -> Lot {
+    let payload = &qr[OPENPGP4FPR_SCHEME.len()..];
+
+    let (fingerprint, fragment) = match payload.find('#').map(|offset| {
+        let (fp, rest) = payload.split_at(offset);
+        // need to remove the # from the fragment
+        (fp, &rest[1..])
+    }) {
+        Some(pair) => pair,
+        None => return format_err!("Invalid OPENPGP4FPR found").into(),
+    };
+
+    dbg!(fingerprint);
+    dbg!(fragment);
+
+    // replace & with \n to match expected param format
+    let fragment = fragment.replace('&', "\n");
+    dbg!(&fragment);
+
+    // Then parse the parameters
+    let param: Params = match fragment.parse() {
+        Ok(params) => params,
+        Err(err) => return err.into(),
+    };
+    dbg!(&param);
+
+    let addr = if let Some(addr) = param.get(Param::Forwarded) {
+        match normalize_address(addr) {
+            Ok(addr) => addr,
+            Err(err) => return err.into(),
+        }
+    } else {
+        return format_err!("Missing address").into();
+    };
+
+    // what is up with that param name?
+    let name = if let Some(encoded_name) = param.get(Param::SetLongitude) {
+        match percent_decode_str(encoded_name).decode_utf8() {
+            Ok(name) => name.to_string(),
+            Err(err) => return format_err!("Invalid name: {}", err).into(),
+        }
+    } else {
+        "".to_string()
+    };
+
+    let invitenumber = param.get(Param::ProfileImage).map(|s| s.to_string());
```

Comment:
that Param::ProfileImage  sounds odd here - in c it was just 'i' which matches what is in the example docstring above.  So maybe it's better to just use 'i' here.

Reply:
seems they are reused for different purposes :/ will think about a better solution